File: membership-token/cli/src/processor/init_selling_resource.rs
//! Module provide handler for `InitSellingResource` command.

use super::UiTransactionInfo;
use crate::error;
use anchor_lang::{InstructionData, ToAccountMetas};
use mpl_membership_token::utils::find_vault_owner_address;
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    instruction::Instruction, pubkey::Pubkey, signature::Signer, signer::keypair::Keypair,
    system_program, sysvar::rent, transaction::Transaction,
};

/// Additional `InitSellingResource` instruction info, that need to be displayed in TUI.
#[derive(Debug)]
pub struct InitSellingResourceUiInfo {
    selling_resource: Keypair,
    vault_owner: Pubkey,
}

impl UiTransactionInfo for InitSellingResourceUiInfo {
    fn print(&self) {
        println!(
            "InitSellingResource::selling_resource - {:?}",
            self.selling_resource
        );
        println!("InitSellingResource::vault_owner - {:?}", self.vault_owner);
    }
}

pub fn init_selling_resource(
    client: &RpcClient,
    payer: &Keypair,
    store: &Pubkey,
    admin_keypair: &Keypair,
    selling_resource_owner: &Pubkey,
    resource_mint: &Pubkey,
    master_edition: &Pubkey,
    vault_keypair: &Keypair,
    resource_token: &Pubkey,
    max_supply: Option<u64>,
) -> Result<(Transaction, Box<dyn UiTransactionInfo>), error::Error> {
    let (vault_owner, vault_owner_bump) = find_vault_owner_address(resource_mint, store);
    let selling_resource = Keypair::new();

    let (_, master_edition_bump) = Pubkey::find_program_address(
        &[
            mpl_token_metadata::state::PREFIX.as_bytes(),
            mpl_token_metadata::id().as_ref(),
            resource_mint.as_ref(),
            mpl_token_metadata::state::EDITION.as_bytes(),
        ],
        &mpl_token_metadata::id(),
    );

    let accounts = mpl_membership_token::accounts::InitSellingResource {
        store: *store,
        admin: admin_keypair.pubkey(),
        selling_resource: selling_resource.pubkey(),
        selling_resource_owner: *selling_resource_owner,
        resource_mint: *resource_mint,
        master_edition: *master_edition,
        vault: vault_keypair.pubkey(),
        owner: vault_owner,
        resource_token: *resource_token,
        rent: rent::id(),
        token_program: spl_token::id(),
        system_program: system_program::id(),
    }
    .to_account_metas(None);

    let data = mpl_membership_token::instruction::InitSellingResource {
        _master_edition_bump: master_edition_bump,
        _vault_owner_bump: vault_owner_bump,
        max_supply,
    }
    .data();

    let instruction = Instruction {
        program_id: mpl_membership_token::id(),
        data,
        accounts,
    };

    let recent_blockhash = client.get_latest_blockhash()?;

    Ok((
        Transaction::new_signed_with_payer(
            &[instruction],
            Some(&payer.pubkey()),
            &[payer, admin_keypair, &selling_resource, vault_keypair],
            recent_blockhash,
        ),
        Box::new(InitSellingResourceUiInfo {
            selling_resource,
            vault_owner,
        }),
    ))
}

File: fixed-price-sale/js/src/instructions/initSellingResource.ts
import * as web3 from '@solana/web3.js';
import * as beet from '@metaplex-foundation/beet';
import * as splToken from '@solana/spl-token';

import { PROGRAM_ID } from '../consts';

export type InitSellingResourceInstructionArgs = {
  masterEditionBump: number;
  vaultOwnerBump: number;
  maxSupply: beet.COption<beet.bignum>;
};
const initSellingResourceStruct = new beet.FixableBeetArgsStruct<
  InitSellingResourceInstructionArgs & {
    instructionDiscriminator: number[];
  }
>(
  [
    ['instructionDiscriminator', beet.uniformFixedSizeArray(beet.u8, 8)],
    ['masterEditionBump', beet.u8],
    ['vaultOwnerBump', beet.u8],
    ['maxSupply', beet.coption(beet.u64)],
  ],
  'InitSellingResourceInstructionArgs',
);
export type InitSellingResourceInstructionAccounts = {
  store: web3.PublicKey;
  admin: web3.PublicKey;
  sellingResource: web3.PublicKey;
  sellingResourceOwner: web3.PublicKey;
  resourceMint: web3.PublicKey;
  masterEdition: web3.PublicKey;
  metadata: web3.PublicKey;
  vault: web3.PublicKey;
  owner: web3.PublicKey;
  resourceToken: web3.PublicKey;
};

const initSellingResourceInstructionDiscriminator = [56, 15, 222, 211, 147, 205, 4, 145];

/**
 * Creates a _InitSellingResource_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 */
export function createInitSellingResourceInstruction(
  accounts: InitSellingResourceInstructionAccounts,
  args: InitSellingResourceInstructionArgs,
) {
  const {
    store,
    admin,
    sellingResource,
    sellingResourceOwner,
    resourceMint,
    masterEdition,
    metadata,
    vault,
    owner,
    resourceToken,
  } = accounts;

  const [data] = initSellingResourceStruct.serialize({
    instructionDiscriminator: initSellingResourceInstructionDiscriminator,
    ...args,
  });
  const keys: web3.AccountMeta[] = [
    {
      pubkey: store,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: admin,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: sellingResource,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: sellingResourceOwner,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: resourceMint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: masterEdition,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: metadata,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: vault,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: owner,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: resourceToken,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: web3.SYSVAR_RENT_PUBKEY,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: splToken.TOKEN_PROGRAM_ID,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: web3.SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
  ];

  const ix = new web3.TransactionInstruction({
    programId: new web3.PublicKey(PROGRAM_ID),
    keys,
    data,
  });
  return ix;
}
